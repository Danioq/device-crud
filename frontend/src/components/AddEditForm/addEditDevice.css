.addEditDeviceContainer {
  position: absolute;
  width: 40vw;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  border: 2px solid blue;
  box-shadow: 0 0 1em black;
  border-radius: 25px;
  padding: 5vh 10vh;
}

.cancelButton {
  margin-left: 1rem;
}

.headerInfo {
  margin-bottom: 1rem;
}

.errorMessage {
  color: red;
  height: 1rem;
  margin-bottom: 0.5rem;
}