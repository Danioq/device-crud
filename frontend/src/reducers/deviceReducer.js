const initialState = {
  devices: []
}

const deviceReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'UPDATE_DEVICES':
      return { ...state, devices: action.payload }
    default:
      return state
  }
}

export default deviceReducer
