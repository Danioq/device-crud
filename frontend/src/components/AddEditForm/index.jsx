import * as React from 'react'
import { Formik } from 'formik'
import { useDispatch } from 'react-redux'
import './addEditDevice.css'
import { PropTypes } from 'prop-types'

const specialCharactersPattern = /[ `!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?~]/

function AddEditForm(props) {
  const dispatch = useDispatch()
  const { device, isEdit } = props

  const Header = () => {
    if (isEdit) {
      return <div className="headerInfo">
        <div>Editing:</div> <b>{device.id}</b>
        </div>
    }
    return <div className="headerInfo">Adding Device</div>
  }
  return <div className="addEditDeviceContainer">
    <Header />
    <Formik
      initialValues={{ name: device?.name || '', description: device?.description || '', disabled: device?.disabled || false }}
      validate={values => {
        const errors = {}
        if (!values.name) {
          errors.name = 'Required'
        }
        if (values.name.match(specialCharactersPattern)) {
          errors.name = 'Name should not contain special characters'
        }
        return errors
      } }
      onSubmit={(values) => {
        if (isEdit) {
          dispatch({ type: 'DEVICE_PATCH_REQUEST', payload: { ...values, id: device.id } })
        } else {
          dispatch({ type: 'DEVICE_POST_REQUEST', payload: { ...values } })
        }
      } }
    >
      {({ values, errors, touched, handleChange, handleBlur, handleSubmit }) => (
        <form onSubmit={handleSubmit}>
          <div className="form-group">
            <label htmlFor="addNameInput">Name</label>
            <input
              type="text"
              name="name"
              onChange={handleChange}
              onBlur={handleBlur}
              value={values.name}
              className="form-control"
              id="addNameInput" />
            <div className="errorMessage">{errors.name && touched.name && errors.name}</div>
            <div className="form-group"></div>
            <label htmlFor="addDescriptionInput">Description</label>
            <textarea
              type="text"
              name="description"
              onChange={handleChange}
              onBlur={handleBlur}
              value={values.description}
              className="form-control"
              id="addDescriptionInput"
              rows="5"
              />
            <div className="form-check">
              <input
                type="checkbox"
                name="disabled"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.disabled}
                className="form-check-input"
                id="addDisabledCheck" />
              <label htmlFor="addDisabledCheck" className="form-check-label">Disabled</label>
            </div>
            <button type="submit" className="btn btn-primary">Add Device</button>
            <button className="btn btn-primary cancelButton" onClick={() => history.goBack()}>Cancel</button>
          </div>
        </form>
      )}
    </Formik>
  </div>
}

AddEditForm.propTypes = {
  device: PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.string,
    description: PropTypes.string,
    disabled: PropTypes.bool
  }),
  isEdit: PropTypes.bool
}

export default AddEditForm
