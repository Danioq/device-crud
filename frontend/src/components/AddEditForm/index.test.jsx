/* eslint-disable no-undef */
import * as React from 'react'
import renderer from 'react-test-renderer'
import AddEditForm from '.'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import deviceReducer from './../../reducers/deviceReducer'

const store = createStore(deviceReducer)
const device = {
  id: '09458915-dfbc-4408-a7c4-5d3162f43964',
  name: 'Device Name',
  description: 'This is device description',
  disabled: false
}

test('Device Match Snapshot', () => {
  const component = renderer.create(<Provider store={store}><AddEditForm device={device} isEdit/></Provider>)
  expect(component).toMatchSnapshot()
})
