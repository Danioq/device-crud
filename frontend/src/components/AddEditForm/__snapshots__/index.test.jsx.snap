// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Device Match Snapshot 1`] = `
<div
  className="addEditDeviceContainer"
>
  <div
    className="headerInfo"
  >
    <div>
      Editing:
    </div>
     
    <b>
      09458915-dfbc-4408-a7c4-5d3162f43964
    </b>
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="addNameInput"
      >
        Name
      </label>
      <input
        className="form-control"
        id="addNameInput"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="Device Name"
      />
      <div
        className="form-group"
      />
      <label
        htmlFor="addDescriptionInput"
      >
        Description
      </label>
      <textarea
        className="form-control"
        id="addDescriptionInput"
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        rows="5"
        type="text"
        value="This is device description"
      />
      <div
        className="form-check"
      >
        <input
          className="form-check-input"
          id="addDisabledCheck"
          name="disabled"
          onBlur={[Function]}
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
        <label
          className="form-check-label"
          htmlFor="addDisabledCheck"
        >
          Disabled
        </label>
      </div>
      <button
        className="btn btn-primary"
        type="submit"
      >
        Add Device
      </button>
      <button
        className="btn btn-primary cancelButton"
        onClick={[Function]}
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
